src/transformers/models/whisper/generation_whisper.py

```diff
@@ -1135,81 +1157,83 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
         generation_config.forced_decoder_ids = None
 
         is_lang_id_undefined = len(init_tokens) <= 1 or (len(init_tokens) > 1 and init_tokens[1] is None)
-        if language is not None:
-            if language in generation_config.lang_to_id.keys():
-                language_token = language
-            elif language in TO_LANGUAGE_CODE.keys():
-                language_token = f"<|{TO_LANGUAGE_CODE[language]}|>"
-            elif language in TO_LANGUAGE_CODE.values():
-                language_token = f"<|{language}|>"
-            else:
-                is_language_code = len(language) == 2
-                raise ValueError(
-                    f"Unsupported language: {language}. Language should be one of:"
-                    f" {list(TO_LANGUAGE_CODE.values()) if is_language_code else list(TO_LANGUAGE_CODE.keys())}."
+
+        # Make sure language is a list of strings of the correct length
+        if isinstance(language, (list, tuple)):
+            if any(l is None for l in language):
+                raise TypeError(
+                    "Expected `language` to be `None`, a single string (e.g. `'en'`), or a list of strings with length equal to the batch size (e.g. `('en', 'fr')` for a batch size of 2). Got a list containing `None`."
                 )
-            if language_token not in generation_config.lang_to_id:
+            if len(language) != batch_size:
                 raise ValueError(
-                    f"{language_token} is not supported by this specific model as it is not in the `generation_config.lang_to_id`."
-                    "(You should just add it to the generation config)"
+                    "When passing a list of languages, the length of the list must match the batch size. "
+                    f"Expected length of {batch_size}, but got {len(language)} languages."
                 )
+            languages = language
+        elif language is None:
+            # Language will be detected for each item in batch
+            languages = [None] * batch_size
+        else:
+            languages = [language]  # Use a length-1 list now, broadcast later
 
-            lang_id = generation_config.lang_to_id[language_token]
+        # Separate init_tokens for each language
+        init_tokens = [copy.copy(init_tokens) for _ in languages]
 
-            # if language is defined it'll overwrite language ids that might have already been defined via the generation_config
-            replace_or_add(init_tokens, lang_id, generation_config.lang_to_id.values())
```

Comment:
I'm not sure why this function was used here. The language token is always `init_tokens[1]`, so it's doing the same thing as this bit in the `elif` branch (only slower): https://github.com/huggingface/transformers/blob/56b64bf1a51e29046bb3f8ca15839ff4d6a92c74/src/transformers/models/whisper/generation_whisper.py#L1177-L1181
So I unified it.